### PR TITLE
PostAgent: allow sending arbitrary string data

### PR DESCRIPTION
**Pull request by dsander**
When `content_type` contains a MIME type and `payload` is not a Hash the string in `payload` will be send as the HTTP body and `content_type` is send as the `Content-Type` header.

#1361

### CODE REVIEW

File: app/models/agents/post_agent.rb [hunk 1]
```diff
@@ -13,7 +15,13 @@ class PostAgent < Agent
...
       The `method` used can be any of `get`, `post`, `put`, `patch`, and `delete`.
 
-      By default, non-GETs will be sent with form encoding (`application/x-www-form-urlencoded`).  Change `content_type` to `json` to send JSON instead.  Change `content_type` to `xml` to send XML, where the name of the root element may be specified using `xml_root`, defaulting to `post`.
+      By default, non-GETs will be sent with form encoding (`application/x-www-form-urlencoded`).
+
+      Change `content_type` to `json` to send JSON instead.
+
+      Change `content_type` to `xml` to send XML, where the name of the root element may be specified using `xml_root`, defaulting to `post`.
+
+      When `content_type` containts a [MIME](https://en.wikipedia.org/wiki/Media_type) type, the interpolated value of `payload` will be send as a string in the HTTP body message and `content_type` will be set as the `Content-Type` HTTP header.
```
Comment (cantino) [inline app/models/agents/post_agent.rb]: Minor edit:

```
When `content_type` contains a [MIME](https://en.wikipedia.org/wiki/Media_type) type, and `payload` is a string, its interpolated value will be sent as a string in the HTTP request's body and the request's `Content-Type` HTTP header will be set to `content_type`.
```

File: app/models/agents/post_agent.rb [hunk 2]
```diff
@@ -66,10 +75,19 @@ def validate_options
         errors.add(:base, "post_url and expected_receive_period_in_days are required fields")
       end
 
-      if options['payload'].present? && !options['payload'].is_a?(Hash)
+      if options['payload'].present? && %w[get delete].include?(method) && !options['payload'].is_a?(Hash)
```
Comment (cantino) [inline app/models/agents/post_agent.rb]: `head` too?  Or do we not support those?

File: app/models/agents/post_agent.rb [hunk 3]
```diff
@@ -66,10 +75,19 @@ def validate_options
...
         errors.add(:base, "if provided, payload must be a hash")
       end
 
+      if options['payload'].present? && %w[post put patch].include?(method)
+        if !(options['content_type'] =~ MIME_RE || options['payload'].is_a?(Hash))
```
Comment (cantino) [inline app/models/agents/post_agent.rb]: Is this more readable? 

`if !options['payload'].is_a?(Hash) && options['content_type'] !~ MIME_RE`

File: app/models/agents/post_agent.rb (hunk 1 above)
Comment (cantino) [inline app/models/agents/post_agent.rb]: Maybe mention `no_merge` too?

**Comment by cantino** [comment]
Nice!

File: app/models/agents/post_agent.rb (hunk 3 above)
Comment (dsander) [inline app/models/agents/post_agent.rb]: It is! That was a remnant of a previous version in which I tried to put it all on online conditional :)

File: app/models/agents/post_agent.rb (hunk 2 above)
Comment (dsander) [inline app/models/agents/post_agent.rb]: We do not support those at the moment.

File: app/models/agents/post_agent.rb (hunk 1 above)
Comment (dsander) [inline app/models/agents/post_agent.rb]: Thanks, done.

**Comment by dsander** [comment]
Thanks for the review.

**Pull request merged**